jumanji/environments/packing/bin_pack/env.py

```diff
@@ -701,3 +714,271 @@ def inclusion_check(ems: EMS, ems_mask: chex.Array) -> chex.Array:
+        return (
+            super()
+            .observation_spec()
+            .replace(
+                items=items,
+                items_mask=items_mask,
+                items_placed=items_placed,
+                action_mask=action_mask,
+            )
+        )
+
+    def action_spec(self) -> specs.MultiDiscreteArray:
+        """Specifications of the action expected by the `BinPack` environment.
+
+        Returns:
+            MultiDiscreteArray (int32) of shape (obs_num_ems, max_num_items).
+            - ems_id: int between 0 and obs_num_ems - 1 (included).
+            - item_id: int between 0 and max_num_items - 1 (included).
+        """
+        num_values = jnp.array(
+            [6, self.obs_num_ems, self.generator.max_num_items], jnp.int32
+        )
+        return specs.MultiDiscreteArray(num_values=num_values, name="action")
+
+    def step(
+        self, state: State, action: chex.Array
+    ) -> Tuple[State, TimeStep[Observation]]:
+        """Run one timestep of the environment's dynamics. If the action is invalid, the state
+        is not updated, i.e. the action is not taken, and the episode terminates.
+
+        Args:
+            state: `State` object containing the data of the current instance.
+            action: jax array (int32) of shape (2,): (ems_id, item_id). This means placing the given
+                item at the location of the given EMS. If the action is not valid, the flag
+                `invalid_action` will be set to True in `timestep.extras` and the episode
+                terminates.
+
+        Returns:
+            state: `State` object corresponding to the next state of the environment.
+            timestep: `TimeStep` object corresponding to the timestep returned by the environment.
+                Also contains metrics in the `extras` field:
+                - volume_utilization: utilization (in [0, 1]) of the container.
+                - packed_items: number of items that are packed in the container.
+                - ratio_packed_items: ratio (in [0, 1]) of items that are packed in the container.
+                - active_ems: number of EMSs in the current instance.
+                - invalid_action: True if the action that was just taken was invalid.
+                - invalid_ems_from_env (optional): True if the environment produced an EMS that was
+                    invalid. Only available in debug mode.
+        """
+        action_is_valid = state.action_mask[tuple(action)]  # type: ignore
+
+        orientation, obs_ems_id, item_id = action
+        ems_id = state.sorted_ems_indexes[obs_ems_id]
+
+        # Pack the item if the provided action is valid.
+        next_state = jax.lax.cond(
+            action_is_valid,
+            lambda s: self._pack_item(s, ems_id, item_id, orientation),
+            lambda s: s,
+            state,
+        )
+        # Make the observation.
+        next_state, observation, extras = self._make_observation_and_extras(next_state)
```

Comment:
I think it would be cleaner to make the observation flat before passing it to the network (have a flat spec, and flatten it in make_observation_and_extras.

Reply:
Done ✅